[16.0][ADD] volunteer: add shift generator and recurrent subscription models

aydrpm opened the pull request:
## Description

This PR introduces the foundation for recurring shift management:

- Add initial models for recurrent shift generator and subscription, with basic views and access rights
- Use mixins to share common logic between generator and shift models
- Add new "Registration" parent menu with submenus for "Participation" and "Subscription" (admin-only access)
- Add configuration settings for shift_nb_occurrence in global settings


## Odoo task (if applicable)



## Checklist before approval

- [ ] Tests are present (or not needed).
- [ ] Credits/copyright have been changed correctly.
- [ ] Change log snippet is present.
- [ ] (If a new module) Moving this to OCA has been considered.

aydrpm committed:
[ADD] Shift mixin

aydrpm committed:
[FIX] Participation: prevent ensure_one() error

Fix ensure_one() error in check_remaining_slot when calling
get_booking_status()

aydrpm committed:
Add generator/subscription models and access rules

aydrpm committed:
Add generator/subscription views and menus:

- Basic tree views for generator and subscription with generator menu
- Form view for generator
- Menu refactoring with new menu "Registration" and sub-menu
  "Participation" and "Subscription" with admin only access

aydrpm committed:
Add configuration settings for shift_nb_occurrence

aydrpm committed:
Fix test sql constraints with mute logger

remytms commented on volunteer/models/volunteer_shift_recurrent_generator.py:
```diff
@@ -0,0 +1,95 @@
+        elif self.interval_type == "years":
+            return relativedelta(years=self.interval)
+        else:
+            raise ValidationError(_("The interval type is not valid."))
```
I think that here using `UserError` is better that `ValidationErorr`. See [Odoo error definition](https://github.com/odoo/odoo/blob/16.0/odoo/exceptions.py#L18).

remytms commented on volunteer/models/volunteer_shift_recurrent_subscription.py:
```diff
@@ -0,0 +1,42 @@
+    _inherit = [
+        "mail.thread",
+        "mail.activity.mixin",
+    ]
```
```suggestion
    ]
    
```
Blank line here to separate _ prefixed var than others.

remytms reviewed (changes requested):
Great great job. Some very minor changes requested.

aydrpm committed:
Fix: apply review feedbacks

- add newline and date fields comment for consistency
- use UserError instead of ValidationError

aydrpm commented:
Thanks for the review and the link about errors! All changes applied in commit [9886d11](https://github.com/beescoop/Obeesdoo/pull/564/commits/9886d116a1c09a270cb1e1ac75f53d242474de67)

remytms reviewed (approved):
LGTM.

remytms commented:
/ocabot merge minor
